distsql: basic flow scheduler

RaduBerinde opened the pull request:
Implementing a very basic flow scheduler. Currently it limits the number of
flows run in parallel.

This change also adds a mechanism to time out flows that never get inbound
connections to their streams, and correctly accounts for the inbound streams in
the flow waitgroup (fixes #9985).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10201)

<!-- Reviewable:end -->

RaduBerinde committed:
distsql: basic flow scheduler

Implementing a very basic flow scheduler. Currently it limits the number of
flows run in parallel.

This change also adds a mechanism to time out flows that never get inbound
connections to their streams, and correctly accounts for the inbound streams in
the flow waitgroup (fixes #9985).

andreimatei commented:
LGTM. Good stuff, as always.
Just a random assortment of suggestions and requests for docs.

---

Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql/flow.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2hdg7-y4E36Gjf3ab:-KV2hdg7-y4E36Gjf3ac:btemjl3) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L68)):*
> ```Go
> )
> 
> type inboundStreamInfo struct {
> ```

Do with this what you will:

I think it'd be really useful to hang some high-level comments around here somewhere about how streams get connected to a Flow. Some node calls some RPC, then `flow.connectInboundStream()` gets called, and `ProcessInboundStream` is the glue between the grpc stream and this `RowReceiver.

Also I think it's worth mentioning that the `receiver` is an InputSync. It is confusing (to me) that this ostensibly "inbound" structure contains a `RowReceiver` - an outbound interface. Something explaining that the reason for this `inboundStream` structure is actually not so it can be used by the other components of the flow (i.e. the processors), but it's for the rendez-vous with the other connected flows from the other nodes, would help.

---

*[pkg/sql/distsql/flow.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2tDgnDgrquPzrA0o7:-KV2tDgnDgrquPzrA0o8:bo8mtjb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L68)):*
> ```Go
> )
> 
> type inboundStreamInfo struct {
> ```

how about having this `inboundStreamInfo` have a `Close()` method that decrements the flow's `waitGroup`? It would be called from both the timer routine and `ProcessInboundStream`, and so we don't need to pass the `waitGroup` to `ProcessInboundStream`. Seems like better encapsulation of the cleanup logic.

---

*[pkg/sql/distsql/flow.go, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2cnFq1mflxhneSjC2:-KV2cnFrSq_fWh5v_PfC:bz1q4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L80)):*
> ```Go
> 
> 	flowRegistry       *flowRegistry
> 	simpleFlowConsumer RowReceiver
> ```

mind documenting `simpleFlowConsumer`?  Is it mutually exclusive with `outboxes`?  If so, can you move it next them?

---

*[pkg/sql/distsql/flow.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2sDGEmRt2MbgQSS_w:-KV2sDGEmRt2MbgQSS_x:b38l81k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L81)):*
> ```Go
> 	flowRegistry       *flowRegistry
> 	simpleFlowConsumer RowReceiver
> 	waitGroup          sync.WaitGroup
> ```

mind documenting this `waitGroup`? Particularly now that all sorts of different beasts register with it.

---

*[pkg/sql/distsql/flow.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2rPI5WaPuAApWahOA:-KV2rPI5WaPuAApWahOB:b4dxamc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L87)):*
> ```Go
> 	localStreams map[LocalStreamID]RowReceiver
> 
> 	doneChan chan<- *Flow
> ```

lol what's this? The flow is supposed to send itself on this chan when it's closed? Seems a bit bizarre to me :)
At lease document, or maybe we can take a cleanup closure instead? Or a `FlowEventListener` interface that'd be more self-documenting.

... After reading more code - do we even use the `*Flow` that's sent on the chan? Can we get rid of it?

---

*[pkg/sql/distsql/flow.go, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2kMg6v2NnddLWHRLW:-KV2kMg7XB8RLlRTpCU6:bec968r) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L89)):*
> ```Go
> 	doneChan chan<- *Flow
> 
> 	mu struct {
> ```

might be worth documenting the point of this new mutex - I assume it's serializing access between the timer and `connectInboundStream`. And I guess an important  point is that using the `receiver`s inside the `StreamInfo`s doesn't need to be protected.

---

*[pkg/sql/distsql/flow_scheduler.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2w_aQpSBpzAtDCKhW:-KV2w_aQpSBpzAtDCKhX:buc27ne) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L29)):*
> ```Go
> const maxRunningFlows = 500
> 
> type flowScheduler struct {
> ```

At least give a brother a sentence about this type. It's gonna be a workhorse.

---

*[pkg/sql/distsql/flow_scheduler.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2vFpmnIewgyztzK3-:-KV2vFpnusxCdWrTo2ck:b-j0o2aw) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L45)):*
> ```Go
> 		AmbientContext: ambient,
> 		stopper:        stopper,
> 		flowDoneCh:     make(chan *Flow, 1),
> ```

why `1` buffer, and not more? Like... 10.

---

*[pkg/sql/distsql/flow_scheduler.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2vmo-dDWckDHWm4Ec:-KV2vmo-dDWckDHWm4Ed:b-sxf900) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L52)):*
> ```Go
> 
> func (fs *flowScheduler) canRunFlow(_ *Flow) bool {
> 	// TODO(radu): we will have more complex resource accounting (like memory).
> ```

putting this on @knz's radar

---

*[pkg/sql/distsql/flow_scheduler.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2w5wCP2uqeMlLR7dw:-KV2w5wCP2uqeMlLR7dx:ba41dci) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L57)):*
> ```Go
> }
> 
> func (fs *flowScheduler) runFlowNow(f *Flow) {
> ```

please document that this doesn't block

---

*[pkg/sql/distsql/flow_scheduler.go, line 92 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2uw5ZkRmqEJKsF0Ut:-KV2uw5ZkRmqEJKsF0Uu:bvbi75c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L92)):*
> ```Go
> 		for {
> 			if stopped && fs.mu.numRunning == 0 {
> 				// TODO(radu): somehow error out the flows that are still in the queue.
> ```

close (one of) their input channels?

---

*[pkg/sql/distsql/inbound.go, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2wsWpCziBQhciXLMG:-KV2wsWq_Kyssh8uAtTl:b63evgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/inbound.go#L35)):*
> ```Go
> 	firstMsg *StreamMessage,
> 	dst RowReceiver,
> 	wg *sync.WaitGroup,
> ```

see comment above about maybe taking a StreamInfo instead of the RowReceiver, and getting rid of `wg`.

---

*[pkg/sql/distsql/server.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2xGliLTrQYb5rnBNF:-KV2xGliLTrQYb5rnBNG:b-8845cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/server.go#L137)):*
> ```Go
> 
> 	if err := ds.Stopper.RunTask(func() {
> 		f.waitGroup.Add(1)
> ```

can we move manipulation of the `waitgroup` completely inside `mbox`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-:-KV2xOeKK_gqfHQh3D5_:b-54lxq4)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql/flow_scheduler.go, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2vmo-dDWckDHWm4Ec:-KV4Tq1ChwCBAgZj9tf-:but32l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L52)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> putting this on @knz's radar
</details>
Thanks, duly noted.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR! Sorry for skimping on docs, I planned to go through it again and add some but forgot.

---

Reviewed 2 of 7 files at r1.
Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/sql/distsql/flow.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2hdg7-y4E36Gjf3ab:-KV4qTtN1vR-iIT3Y-9V:b-ld40zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L68)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Do with this what you will:
> 
> I think it'd be really useful to hang some high-level comments around here somewhere about how streams get connected to a Flow. Some node calls some RPC, then `flow.connectInboundStream()` gets called, and `ProcessInboundStream` is the glue between the grpc stream and this `RowReceiver.
> 
> Also I think it's worth mentioning that the `receiver` is an InputSync. It is confusing (to me) that this ostensibly "inbound" structure contains a `RowReceiver` - an outbound interface. Something explaining that the reason for this `inboundStream` structure is actually not so it can be used by the other components of the flow (i.e. the processors), but it's for the rendez-vous with the other connected flows from the other nodes, would help.
</details>
Done.  RowReceiver represents an entity that receives rows, so it's an inbound interface.. If you have suggestions for better names let me know.

---

*[pkg/sql/distsql/flow.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2tDgnDgrquPzrA0o7:-KV4tMIVWS6XRlHfKdSb:b-8o4kt4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L68)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> how about having this `inboundStreamInfo` have a `Close()` method that decrements the flow's `waitGroup`? It would be called from both the timer routine and `ProcessInboundStream`, and so we don't need to pass the `waitGroup` to `ProcessInboundStream`. Seems like better encapsulation of the cleanup logic.
</details>
Good idea. I added it as a method on the flow though.

---

*[pkg/sql/distsql/flow.go, line 80 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2cnFq1mflxhneSjC2:-KV4wHTgl585jzs7lvgi:b-w8emup) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L80)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> mind documenting `simpleFlowConsumer`?  Is it mutually exclusive with `outboxes`?  If so, can you move it next them?
</details>
Done. It's not necessarily mutually exclusive though it's hard to think of an example where we would want both.

---

*[pkg/sql/distsql/flow.go, line 81 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2sDGEmRt2MbgQSS_w:-KV4wZrg1QeMfdDKFOKY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L81)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> mind documenting this `waitGroup`? Particularly now that all sorts of different beasts register with it.
</details>
Done.

---

*[pkg/sql/distsql/flow.go, line 87 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2rPI5WaPuAApWahOA:-KV4w_VqExFbaWg3xeq0:bkeqe4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L87)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> lol what's this? The flow is supposed to send itself on this chan when it's closed? Seems a bit bizarre to me :)
> At lease document, or maybe we can take a cleanup closure instead? Or a `FlowEventListener` interface that'd be more self-documenting.
> 
> ... After reading more code - do we even use the `*Flow` that's sent on the chan? Can we get rid of it?
</details>
In general, the flow scheduler will need to know which flow completed. Currently it's so simple that it doesn't care but it will. Changed it to a closure though.

---

*[pkg/sql/distsql/flow.go, line 89 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2kMg6v2NnddLWHRLW:-KV4yvu06SlfckpY6bDN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L89)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> might be worth documenting the point of this new mutex - I assume it's serializing access between the timer and `connectInboundStream`. And I guess an important  point is that using the `receiver`s inside the `StreamInfo`s doesn't need to be protected.
</details>
Done.

---

*[pkg/sql/distsql/flow_scheduler.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2w_aQpSBpzAtDCKhW:-KV4zPF9EBciT_cwNU6b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L29)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> At least give a brother a sentence about this type. It's gonna be a workhorse.
</details>
Done.

---

*[pkg/sql/distsql/flow_scheduler.go, line 45 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2vFpmnIewgyztzK3-:-KV4zWmz2cl-i4WX9ZoD:brws98h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L45)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> why `1` buffer, and not more? Like... 10.
</details>
8. Just because I am my own man!!!

---

*[pkg/sql/distsql/flow_scheduler.go, line 57 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2w5wCP2uqeMlLR7dw:-KV4z8LzXnCNy3Vft4bs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L57)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> please document that this doesn't block
</details>
Done.

---

*[pkg/sql/distsql/flow_scheduler.go, line 92 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2uw5ZkRmqEJKsF0Ut:-KV4zjlmNAytJX62_h5h:b-c0qdii) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow_scheduler.go#L92)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> close (one of) their input channels?
</details>
These flows aren't even started. We would need to go to their outboxes and send out some errors. This may tie in with connecting outboxes early which we were talking about.

---

*[pkg/sql/distsql/inbound.go, line 35 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2wsWpCziBQhciXLMG:-KV5-0YaibgkHeW7NnqZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/inbound.go#L35)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> see comment above about maybe taking a StreamInfo instead of the RowReceiver, and getting rid of `wg`.
</details>
Done.

---

*[pkg/sql/distsql/server.go, line 137 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2xGliLTrQYb5rnBNF:-KV5-C2SpUpkutXohhcI:b-hmu6mt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/server.go#L137)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> can we move manipulation of the `waitgroup` completely inside `mbox`?
</details>
Right now `outbox` does not depend on `Flow` and I think that helps from a code modularity/understanding perspective (and note that it's the same type we use for regular outboxes which have a slightly different logic around the Add).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-:-KV5-k0TQvALsyaIJ9ki:b-g3y3hs)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Btw should I `s/inboundStreamInfo/inboundEndpoint`? (or something else?)

---

Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-:-KV54Bm7GxqNSy-fNQOZ:bvm1pv5)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
LGTM

I think `inboundStreamInfo` is fine for the type, but maybe the member of the flow using it should be something like `inboundRendezvousSlots`. And maybe, instead of having the "connected/not connected" states that you've just introduced, a cleaner thing would be to have consumers extract the `rendezvousSlot` they want to connect a stream to from the collection. And then, after the timeout, if there's still slots in that struct, then you know those have not yet been connected.
As I'm also trying to say below where I'm bikeshedding on the comment, I think it's easy to get confused about the flow of data, so if we can make it suggestive that the flow doesn't actually use these "streamInfos", I think it'd be a good thing. So maybe I should take my previous suggestion further - maybe these "streamInfos" should not be hanging from the flow, but from a higher-level orchestrator - maybe the new scheduler.

Also feel free to ignore all of this.

---

Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql/flow.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2hdg7-y4E36Gjf3ab:-KV6koSB4E2IHKteVzM3:b-ahlg8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L68)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Done.  RowReceiver represents an entity that receives rows, so it's an inbound interface.. If you have suggestions for better names let me know.
</details>
well this `receiver` is indeed inbound, but that's because it's being consumed by the processors in this flow *through another interface* which is not reflected in this struct. People that use it as a *RowReceiver* are pushing rows into it, so they use it as an *outbound* structure. That's the point that I think it would be valuable to make here. Otherwise I think there's a bit of a disconnect between the `inbound` name of the structure and the type used here. And also any comment that would help me better understand how this ties to/is consumed by the other components of the flow, the better.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-:-KV6kbexRCsodvRkx1WC:b-lymp19)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I moved the inbound stream connection to FlowRegistry. I agree it makes more sense here.

---

Review status: 0 of 8 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsql/flow.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2hdg7-y4E36Gjf3ab:-KVQawoDXeYcYqw581Iz:b9utmvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L68)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> well this `receiver` is indeed inbound, but that's because it's being consumed by the processors in this flow *through another interface* which is not reflected in this struct. People that use it as a *RowReceiver* are pushing rows into it, so they use it as an *outbound* structure. That's the point that I think it would be valuable to make here. Otherwise I think there's a bit of a disconnect between the `inbound` name of the structure and the type used here. And also any comment that would help me better understand how this ties to/is consumed by the other components of the flow, the better.
</details>
Through which "another interface" is the receiver consumed? Processors don't "pull" rows from this entity through some other interface. The rows get pushed to the processor through the `RowReceiver` interface.

I understand what you are saying - yes receiver is an "outbound" interface in that we push rows to it. But this thing as a whole is a conceptual inbound point for the flow itself (inbound refers to something higher-level than the particular mechanism through which we happen to move rows).

I'm not sure how to make that more clear. I updated the comment a bit. If you have a specific suggestion on how to improve it let me know.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-:-KVQgo4KYbZzClkEZVv6:bxayyp1)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 8 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsql/flow.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-KV2hdg7-y4E36Gjf3ab:-KVaSTF6AmnBzbRyUOzy:b-mchujp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f657d84339597281bdf7ccb6b0adddbdfd74c35f/pkg/sql/distsql/flow.go#L68)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Through which "another interface" is the receiver consumed? Processors don't "pull" rows from this entity through some other interface. The rows get pushed to the processor through the `RowReceiver` interface.
> 
> I understand what you are saying - yes receiver is an "outbound" interface in that we push rows to it. But this thing as a whole is a conceptual inbound point for the flow itself (inbound refers to something higher-level than the particular mechanism through which we happen to move rows).
> 
> I'm not sure how to make that more clear. I updated the comment a bit. If you have a specific suggestion on how to improve it let me know.
> 
> 
</details>
as discussed, say that "receiver" is a pointer to a processor.
And also say that the collection of `inboundStreams` will be passed to the registry.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10201#-:-KVaSww6FdOOuXx8Me_S:bnfp4nl)*
<!-- Sent from Reviewable.io -->
